sklearn/utils/_array_api.py

```diff
@@ -669,6 +669,36 @@ def _average(a, axis=None, weights=None, normalize=True, xp=None):
     return sum_ / scale
 
 
+def _median(X, axis=None, xp=None):
+    xp, _ = get_namespace(X, xp=xp)
+
+    if _is_numpy_namespace(xp):
+        return numpy.median(X, axis=axis)
+
+    if X.ndim == 0:
+        return float(X)
+
+    if axis is None:
+        X = xp.reshape(X, (-1,))
+        axis = 0
+
+    X_sorted = xp.sort(X, axis=axis)
+    indexer = [slice(None)] * X.ndim
+    index = X.shape[axis] // 2
+    if X.shape[axis] % 2 == 1:
+        # index with slice to allow mean (below) to work
+        indexer[axis] = slice(index, index + 1)
+    else:
+        indexer[axis] = slice(index - 1, index + 1)
+    indexer = tuple(indexer)
+
+    # Use mean in both odd and even case to coerce data type,
+    # using out array if needed.
+    rout = xp.mean(X_sorted[indexer], axis=axis)
```

Comment:
Technically the spec states that NaNs are propagated (https://data-apis.org/array-api/latest/API_specification/generated/array_api.mean.html#mean) but there is also a note that says:

> Array libraries, such as NumPy, PyTorch, and JAX, currently deviate from this specification in their handling of components which are NaN when computing the arithmetic mean.